# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline


In [2]:
os.getcwd()

'C:\\Users\\Pranay\\Projects\\GitHub\\Guided-Capstone'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df=pd.read_csv('data/step3_output.csv')
df.head()

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          1     1       3  ...   
4      0        0.0          0          0     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0             150.0       36.0   
1              800.0            80.0             115.0       49.0   
2               88.0             0.0             110.0       84.0   
3              380.0           200.0             150.0       54.0   
4              862.0             0.0             150.0       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0          34.0              152.0   
1            250.0          74.0          78.0              104.0   
2            300.0          47.0          47.0              107.0   
3            400.0          49.0          47.0              150.0   
4            350.0          78.0          78.0              140.0   

   NightSkiing_ac  clusters  
0            30.0         2  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:

#dfo=df['state']
#df.select_dtypes(include=['object']) # select object type columns
#df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)
#df.head()

In [5]:
dfo=pd.get_dummies(df['state'])

In [6]:
df = pd.concat([df, dfo], axis=1)

In [7]:
df.drop('state',axis=1,inplace=True)
df.head()

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          1     1       3       1  ...             0   
4        0.0          0          0     1       2       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [8]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
Y= df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [9]:
X_scaled

array([[-0.48369489, -0.9313    ,  0.        , ..., -0.13284223,
        -0.26057865, -0.17251639],
       [ 2.32980236,  1.64949964,  0.        , ..., -0.13284223,
        -0.26057865, -0.17251639],
       [ 1.29933167, -0.40691442,  0.        , ..., -0.13284223,
        -0.26057865, -0.17251639],
       ...,
       [ 1.30182978,  1.25706729,  0.        , ..., -0.13284223,
        -0.26057865,  5.7965507 ],
       [ 1.88107921,  0.26142015,  0.        , ..., -0.13284223,
        -0.26057865,  5.7965507 ],
       [ 1.8301802 ,  0.44992477,  0.        , ..., -0.13284223,
        -0.26057865,  5.7965507 ]])

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [10]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
Y = Y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, Y_train, Y_test = train_test_split(X_scaled, Y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [11]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,Y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [14]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [0]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [15]:
from sklearn.metrics import explained_variance_score
explained_variance_score(Y_test, y_pred)

-1.5055313012682492e+23

In [16]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(Y_test, y_pred)

1484020628040.273

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [17]:
lm.intercept_

-377438772391.0725

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [18]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
summit_elev        1.158181e+00
vertical_drop      4.091212e+00
trams              5.675426e+12
fastEight          7.133154e+13
fastSixes          9.113506e+12
fastQuads          2.499858e+12
quad               2.993461e+12
triple             5.268531e+12
double             6.318104e+12
surface            6.137093e+12
total_chairs       1.179439e+13
Runs               3.818359e-01
TerrainParks       3.242188e-01
LongestRun_mi      1.113281e-01
SkiableTerrain_ac  1.078979e+00
Snow Making_ac     1.402344e+00
daysOpenLastYear   1.167969e+00
yearsOpen          8.300781e-01
averageSnowfall    7.460938e-01
AdultWeekday       8.721680e+00
projectedDaysOpen  4.716797e-01
NightSkiing_ac     1.407227e+00
clusters           2.341797e+00
Alaska             2.475124e+12
Arizona            2.475124e+12
California         8.295382e+12
Colorado           4.906888e+12
Connecticut        5.469813e+12
Idaho              5.974020e+12
Illinois           4.262044e+12
Indiana            3.490164e+12
Iowa               4.262044e+12
Maine              5.469813e+12
Maryland           1.857098e+08
Massachusetts      4.906888e+12
Michigan           8.904210e+12
Minnesota          6.856773e+12
Missouri           3.490164e+12
Montana            4.262044e+12
Nevada             2.475124e+12
New Hampshire      7.966850e+12
New Jersey         2.475124e+12
New Mexico         5.974020e+12
New York           1.043983e+13
North Carolina     4.906888e+12
Ohio               4.906888e+12
Oregon             4.262044e+12
Pennsylvania       7.966850e+12
Rhode Island       3.426848e+01
South Dakota       2.475124e+12
Tennessee          2.475124e+12
Utah               4.262044e+12
Vermont            6.856773e+12
Virginia           3.490164e+12
Washington         4.262044e+12
West Virginia      4.262044e+12
Wisconsin          7.966850e+12
Wyoming            5.469813e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [20]:
df.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'],
      dtype='object')

In [21]:
p=['Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming']

In [22]:
df.drop(p,axis=1,inplace=True)
df.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters'],
      dtype='object')

In [23]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
Y= df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [24]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
Y = Y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, Y_train, Y_test = train_test_split(X_scaled, Y, test_size=0.25, random_state=1)

In [25]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,Y_train)

In [26]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [27]:
from sklearn.metrics import explained_variance_score
explained_variance_score(Y_test, y_pred)

0.7443191946285843

In [28]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(Y_test, y_pred)

7.019898494868436

In [29]:
lm.intercept_

56.69057215798981

In [30]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
summit_elev        4.181640e+00
vertical_drop      4.141962e+00
trams              0.000000e+00
fastEight          1.110223e-15
fastSixes          1.665335e-16
fastQuads          9.735145e-01
quad               1.104290e+00
triple             1.432080e+00
double             6.761590e-01
surface            8.159449e-01
total_chairs       6.512819e-01
Runs               4.213733e-01
TerrainParks       8.261410e-01
LongestRun_mi      7.246541e-01
SkiableTerrain_ac  1.328913e+00
Snow Making_ac     1.971707e+00
daysOpenLastYear   1.759572e+00
yearsOpen          1.344808e-01
averageSnowfall    1.684788e+00
AdultWeekday       9.835760e+00
projectedDaysOpen  1.142772e-01
NightSkiing_ac     1.833120e+00
clusters           2.901342e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

In [32]:
df.drop('summit_elev',axis=1,inplace=True)
df.head()

Name  vertical_drop  trams  fastEight  fastSixes  \
0               Hilltop Ski Area            294      0        0.0          0   
1            Sunrise Park Resort           1800      0        0.0          0   
2  Yosemite Ski & Snowboard Area            600      0        0.0          0   
3         Boreal Mountain Resort            500      0        0.0          0   
4                    Dodge Ridge           1600      0        0.0          0   

   fastQuads  quad  triple  double  surface  ...  SkiableTerrain_ac  \
0          0     0       1       0        2  ...               30.0   
1          1     2       3       1        0  ...              800.0   
2          0     0       1       3        1  ...               88.0   
3          1     1       3       1        2  ...              380.0   
4          0     1       2       5        4  ...              862.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0            30.0             150.0       36.0             69.0          30.0   
1            80.0             115.0       49.0            250.0          74.0   
2             0.0             110.0       84.0            300.0          47.0   
3           200.0             150.0       54.0            400.0          49.0   
4             0.0             150.0       69.0            350.0          78.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          34.0              152.0            30.0         2  
1          78.0              104.0            80.0         1  
2          47.0              107.0             0.0         1  
3          47.0              150.0           200.0         1  
4          78.0              140.0             0.0         1  

[5 rows x 24 columns]

#### Model 3

In [33]:
df.columns

Index(['Name', 'vertical_drop', 'trams', 'fastEight', 'fastSixes', 'fastQuads',
       'quad', 'triple', 'double', 'surface', 'total_chairs', 'Runs',
       'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

In [34]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
Y= df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [35]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
Y = Y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, Y_train, Y_test = train_test_split(X_scaled, Y, test_size=0.25, random_state=1)

In [36]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,Y_train)

In [37]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [38]:
from sklearn.metrics import explained_variance_score
explained_variance_score(Y_test, y_pred)

0.7943548039374888

In [39]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(Y_test, y_pred)

6.237503960620418

In [40]:
lm.intercept_

56.583157469797214

In [41]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
vertical_drop      4.228671e+00
trams              0.000000e+00
fastEight          1.776357e-15
fastSixes          3.108624e-15
fastQuads          7.841339e-01
quad               9.412247e-01
triple             1.293720e+00
double             7.489046e-01
surface            4.533923e-01
total_chairs       8.158497e-01
Runs               6.075783e-01
TerrainParks       8.588417e-01
LongestRun_mi      9.072437e-01
SkiableTerrain_ac  2.803883e+00
Snow Making_ac     1.764069e+00
daysOpenLastYear   1.128010e+00
yearsOpen          3.205175e-01
averageSnowfall    2.887246e+00
AdultWeekday       9.736396e+00
projectedDaysOpen  5.749831e-01
NightSkiing_ac     1.327385e+00
clusters           1.941391e+00

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.5055313012682492 | 1484020628040.273 |-|
| Model 2. | 0.7443191946285843  | 7.019898494868436 |'state'|
| Model 3. | 0.7943548039374888  | 6.237503960620418 |'state','summit_elev','base_elev'|

Model Selection:

In [56]:
print("Model 3")

Model 3
